### ISSUE TITLE
Bad warnings when shadowing _ (0.6)

### ISSUE THREAD

**Issue by cstjean** [JuliaLang/julia#20931]
The warnings appear only inside closures.
```julia
julia> for _ in 1:10
           function f(x)
               _ = 10
           end
       end

WARNING: deprecated syntax "_ as an rvalue".

WARNING: deprecated syntax "_ as an rvalue".
```

**Comment by StefanKarpinski** [comment]
Resolved: we're probably not going to be able to fix this in 0.6 but we can add a line number and just tell people to use a different name in these cases.

**Comment by StefanKarpinski** [comment]
Leaving it on 0.6 until the line number is printed.

**Comment by StefanKarpinski** [comment]
Note that this won't be a problem on 1.0 when the deprecation goes away, so it's unclear that a fix beyond reporting a line number is necessary.

**Comment by cstjean** [comment]
It causes typing issues. In the function below, `_` is boxed, allocates 1.5KB for n=100, and takes 10X longer than the same code with one of the `_` replaced with `i`.
```julia
function foo(n)
    s = 0.0
    for _ in 1:n
       function f(x)
           _ = "hey"
       end
        s += 1.0
    end
    return s
end
```
I know that's true for any variable name, but `_` might easily be used multiple times in a function, since its value is supposedly discarded. Maybe that's a separate issue?

**Comment by yuyichao** [comment]
There values are not yet discarded.

**Comment by JeffBezanson** [comment]
#21301

**Comment by StefanKarpinski** [comment]
This may get fixed in the 0.6.x time frame, but since it will go away in 1.0 when this deprecation is removed, it's a fairly low priority. Sorry for the annoyance here, but it's fairly hard to fix and since it's a temporary problem, the effort is a bit hard to justify.
